src/server/cluster/incoming_slot_migration.cc

```diff
@@ -130,16 +139,29 @@ IncomingSlotMigration::IncomingSlotMigration(string source_id, Service* se, Slot
 IncomingSlotMigration::~IncomingSlotMigration() {
 }
 
-bool IncomingSlotMigration::Join() {
-  auto timeout = absl::GetFlag(FLAGS_slot_migration_connection_timeout_ms) * 1ms;
-  if (bc_->WaitFor(timeout)) {
-    state_.store(MigrationState::C_FINISHED);
-    keys_number_ = cluster::GetKeyCount(slots_);
-    return true;
+bool IncomingSlotMigration::Join(long attempt) {
+  const absl::Time start = absl::Now();
+  const absl::Duration timeout =
+      absl::Milliseconds(absl::GetFlag(FLAGS_slot_migration_connection_timeout_ms));
+
+  while (true) {
```

Comment:
why do we have a while loop here? why not just return if 
std::all_of(shard_flows_.begin(), shard_flows_.end(),
                     [&](const auto& flow) { return flow->GetLastAttempt() == attempt; }))
is false?

We will call this join function on each DFLYMIGRATE ACK command sent, how does this works?

Reply:
We are given a `timeout` here, let's say 2s (default).
So we should be returning if (a) we reached the correct join state, or (b) if 2s have passed.
This loop is here to "protect" us from returning earlier than 2s but without (a), in case we were able to join all on an earlier `FIN` (now `LSN`).
So for example if attempt == 4, but one of more of the flows are currently stopped on attempt 3, we will try again, with a timeout equals to the remainder of the time.
I hope this makes sense. If you could phrase some comment to explain that, I'm more than happy to add it!